packages/aws-cdk/lib/api/hotswap/lambda-functions.ts

```diff
@@ -165,8 +182,10 @@ interface CfnDiffTagValue {
 }
 
 interface LambdaFunctionCode {
-  readonly s3Bucket: string;
-  readonly s3Key: string;
+  readonly s3Bucket?: string;
+  readonly s3Key?: string;
+  readonly imageUri?: string;
+  readonly zipFile?: Buffer;
 }
```

Comment:
Not a huge part of this modeling with optional properties. Since this will be purely TypeScript, let's use the full power of union types here:

```ts
type LambdaFunctionCode = S3FunctionCode | DockerFunctionCode;

interface S3FunctionCode {
  readonly s3Bucket: string;
  readonly s3Key: string;
}

interface DockerFunctionCode {
  readonly imageUri: string;
}
```

Reply:
If we use the union type, how do we get values from it when we call updateFunctionCode API?

```ts
      await lambda.updateFunctionCode({
        FunctionName: this.lambdaFunctionResource.physicalName,
        // we need to determine whether resource.code is S3FunctionCode or DockerFunctionCode first
        S3Bucket: resource.code.s3Bucket, // error
        S3Key: resource.code.s3Key, // error
        ImageUri: resource.code.imageUri, // error
      }).promise();
```

Maybe we need to add an additional field like `codeType`...

I'm not a huge fan of optional properties either haha but imho they fit this case well just because `lambda.updateFunctionCode` API is designed as such.

Reply:
We'll need to define a helper function:

```ts
type LambdaFunctionCode = S3FunctionCode | DockerFunctionCode;

interface S3FunctionCode {
  readonly s3Bucket: string;
  readonly s3Key: string;
}

interface DockerFunctionCode {
  readonly imageUri: string;
}

function isS3FunctionCode(code: LambdaFunctionCode): code is S3FunctionCode {
  return (code as any).s3Key !== undefined;
}
```

And then use it like so:

```ts
      const updateFunctionCodePromise = lambda.updateFunctionCode({
        FunctionName: this.lambdaFunctionResource.physicalName,
        S3Bucket: isS3FunctionCode(resource.code) ? resource.code.s3Bucket : undefined,
        S3Key: isS3FunctionCode(resource.code) ? resource.code.s3Key : undefined,
      }).promise();
```

And similarly with `DockerFunctionCode`.

It's a little bit of boilerplate, I agree, but I kind of like the elegance of this solution, vs. making all properties of the interface optional 🙂.

I won't hold the PR on this though if you think this is too much work for something relatively simple.

Reply:
Thanks! I've understood how it works. I tend to think this is a bit too much though, maybe I'll try your technique somewhere in the future!